### ISSUE TITLE
Loss becomes NaN and accuracy stuck at ≈ 0.098 when running EventBasedMNIST_with_log.ipynb

### ISSUE THREAD

**Issue by psarnkaew** [chenxi1228/LeTE#3]
Hello,
I’m very interested in your excellent article and really appreciate your work!
However, I encountered a problem when reproducing the results.
After cloning and running the notebook EventBasedMNIST_with_log.ipynb, both the training and test losses become NaN, and the accuracy stays around 0.098 (≈ 1 / 10) throughout all epochs.

I ran this experiment in Google Colab, and the datasets appeared to download and load correctly.
Here is an example of the output (the complete log is attached in the PDF file ([logs_EventBasedMNIST.pdf](https://github.com/user-attachments/files/22986013/logs_EventBasedMNIST.pdf)). Below is the example of output of the executed experiment.
Epoch [196/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
100%|██████████| 118/118 [00:11<00:00, 10.62it/s]
Epoch [197/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
100%|██████████| 118/118 [00:11<00:00, 10.64it/s]
Epoch [198/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
100%|██████████| 118/118 [00:11<00:00, 10.57it/s]
Epoch [199/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
100%|██████████| 118/118 [00:11<00:00, 10.59it/s]
Epoch [200/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
Final Training Accuracy per epoch: [0.09871666666666666, 0.09871666666666666, 0.09871666666666666, 0.09871666666666666,... 0.09871666666666666, 0.09871666666666666]
Final Training Loss per epoch: [nan, nan, nan,..., nan]
Final Test Accuracy per epoch: [0.098, 0.098, 0.098, 0.098, ..., 0.098]
Final Test Loss per epoch: [nan, nan, nan, ..., nan]

**Comment by chenxi1228** [comment]
> Hello,
> I’m very interested in your excellent article and really appreciate your work!
> However, I encountered a problem when reproducing the results.
> After cloning and running the notebook EventBasedMNIST_with_log.ipynb, both the training and test losses become NaN, and the accuracy stays around 0.098 (≈ 1 / 10) throughout all epochs.
> 
> I ran this experiment in Google Colab, and the datasets appeared to download and load correctly.
> Here is an example of the output (the complete log is attached in the PDF file ([logs_EventBasedMNIST.pdf](https://github.com/user-attachments/files/22986013/logs_EventBasedMNIST.pdf)). Below is the example of output of the executed experiment.
> Epoch [196/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
> 100%|██████████| 118/118 [00:11<00:00, 10.62it/s]
> Epoch [197/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
> 100%|██████████| 118/118 [00:11<00:00, 10.64it/s]
> Epoch [198/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
> 100%|██████████| 118/118 [00:11<00:00, 10.57it/s]
> Epoch [199/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
> 100%|██████████| 118/118 [00:11<00:00, 10.59it/s]
> Epoch [200/200] finished. Train Loss: nan, Train Accuracy: 0.0987, Test Loss: nan, Test Accuracy: 0.0980
> Final Training Accuracy per epoch: [0.09871666666666666, 0.09871666666666666, 0.09871666666666666, 0.09871666666666666,... 0.09871666666666666, 0.09871666666666666]
> Final Training Loss per epoch: [nan, nan, nan,..., nan]
> Final Test Accuracy per epoch: [0.098, 0.098, 0.098, 0.098, ..., 0.098]
> Final Test Loss per epoch: [nan, nan, nan, ..., nan]
> 
> 


You may try using a different random seed. I haven't encountered this issue when running it on EventBasedMNIST, but I have faced it while training other models. This is likely due to some specific implementation details. Also, I recommend trying the EventBasedMNIST.ipynb file instead of the log file, as the latter is only intended for demonstrating logs.

**Comment by psarnkaew** [comment]
I also tested the model on EventBasedMNIST.ipynb using different random seeds (1 and 42), but it produced the same result.

**Comment by psarnkaew** [comment]
I added the following initialization inside the __init__ function of the Spline(nn.Module) class, right after the original definitions of self.base_weight and self.spline_weight:

```
self.base_weight = nn.Parameter(
            torch.randn(self.dim_spline, self.dim_spline) / np.sqrt(self.dim_spline)
        )
self.spline_weight = nn.Parameter(
    torch.randn(self.dim_spline, self.dim_spline, self.grid_size_spline + self.order_spline) /
    (np.sqrt(self.dim_spline) * np.sqrt(self.grid_size_spline + self.order_spline))
)


```
After making this change, the model seems to train properly — the loss and test accuracy are updated as expected, and the results align more closely with those reported in your paper.

However, I’m not entirely sure whether this modification is valid fix for your LeTE or not. Could you please confirm if this is a reasonable fix?

**Comment by chenxi1228** [comment]
From my perspective, yes, your modification is a reasonable fix.

The issue you encountered was very likely caused by the parameter initialization of the spline part. In KAN-based spline modules, it’s common to use a `reset_parameters()` function that initializes the base weights with Kaiming uniform initialization and the spline coefficients with small random noise projected through a curve2coeff function. This helps keep the initial spline functions smooth and prevents exploding gradients or NaN losses.

I believe, your initialization follows a Xavier-style scaling strategy, which controls the variance of activations and gradients. This is sound — if the model now trains stably and reaches expected accuracy, it indicates the initialization properly balances variance across spline coefficients.

As your initialization makes LeTE train stably and produce the expected results, I think it’s fine to use it. The final parameters are learned from data anyway — initialization mainly serves to prevent unstable training or NaN losses in the early stage.

**Comment by chenxi1228** [comment]
> I also tested the model on EventBasedMNIST.ipynb using different random seeds (1 and 42), but it produced the same result.

Using a different random seed simply leads to a different random initialization of the module’s parameters, which can sometimes affect training stability. If your modified initialization consistently eliminates the issue and produces stable results across different random seeds, it is a valid choice to use it.

**Comment by psarnkaew** [comment]
The issue is resolved, so I'm closing it now.
Thank you for your kind reply and your awesome work!

**Comment by chenxi1228** [comment]
> The issue is resolved, so I'm closing it now. Thank you for your kind reply and your awesome work!

Thank you very much for your interest in our work! Hope it can inspire you to create even more wonderful research.
